isb_lib/datacite/__init__.py

```diff
@@ -0,0 +1,129 @@
+import datetime
+import json
+import typing
+import logging
+import requests
+from requests import Response
+from requests.auth import HTTPBasicAuth
+
+from isb_lib.core import parsed_date
+
+CONTENT_TYPE = "application/vnd.api+json"
+DATACITE_URL = "https://api.test.datacite.org"
+
+
+def _dois_headers():
+    headers = {"content-type": CONTENT_TYPE}
+    return headers
+
+
+def _dois_auth(password, username):
+    auth = HTTPBasicAuth(username, password)
+    return auth
+
+
+def _validate_response(response) -> bool:
+    if response.status_code < 200 or response.status_code >= 300:
+        logging.error(
+            "Error requesting new DOI, status code: %s, response %s",
+            response.status_code,
+            str(response.json()),
+        )
+        return False
+    return True
+
+
+def _post_to_datacite(
+    rsession: requests.session, post_data_str: str, username: str, password: str
+) -> Response:
+    response = rsession.post(
+        dois_url(),
+        headers=_dois_headers(),
+        data=post_data_str,
+        auth=_dois_auth(password, username),
+    )
+    return response
+
+
+def _doi_or_none(response: Response) -> typing.Optional[str]:
+    if not _validate_response(response):
+        return None
+    json_response = response.json()
+    draft_id = json_response["data"]["id"]
+    # use the DOI prefix since we creating DOIs with datacite
+    return doi_from_id(draft_id)
+
+
+# See datacite docs here: https://support.datacite.org/docs/api-create-dois
+def datacite_metadata_from_core_record(
+    prefix: str, doi: str, authority: str, core_record: dict
+) -> dict:
```

Comment:
@datadavev I'd really appreciate your eyes on this part.  This is the skeleton of the mapping that we register with datacite.